syntax = "proto3";

option go_package = "./protos";

import "google/protobuf/wrappers.proto";

package library;

service Library {
    // A server-to-client streaming RPC.
    rpc ListBooks(ListBooksRequest) returns (stream Book) {}
    rpc FilterBooks(FilterBooksRequest) returns (stream Book) {}
    rpc SubscribeForBookUpdates(SubscribeForBookUpdatesRequest) returns (stream Book) {}
}

message ListBooksRequest {}

message Book {
    int32 id = 1;
    string title = 2;
    repeated Author authors = 3;
}

message Author {
    int32 id = 1;
    string name = 2;
}

message FilterBooksRequest {
    google.protobuf.StringValue title = 1;
    google.protobuf.StringValue authorName = 2;
}

message SubscribeForBookUpdatesRequest {

}
